scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -0,0 +1,219 @@
+
+(defn kill-task
+  [api-client expected-state-dict pod]
+  (api/kill-task api-client pod)
+  expected-state-dict)
+
+(defn launch-task
+  [api-client expected-state-dict]
+  (api/launch-task api-client expected-state-dict)
+  expected-state-dict)
+
+(defn update-or-delete!
+  "Given a map atom, key, and value, if the value is not nil, set the key to the value in the map. If the value is nil, delete the key entirely."
+  [^IAtom map-atom key value]
+  (if (nil? value)
+    (swap! map-atom dissoc key)
+    (swap! map-atom assoc key value)))
+
+(defn container-status->failure-reason
+  "Maps kubernetes failure reasons to cook failure reasons"
+  [^V1PodStatus pod-status ^V1ContainerStatus status]
+  ; TODO map additional kubernetes failure reasons
+  (let [terminated (-> status .getState .getTerminated)]
+    (cond
+      (= "OutOfMemory" (.getReason pod-status)) :reason-container-limitation-memory
+      (= "Error" (.getReason terminated)) :reason-command-executor-failed
+      (= "OOMKilled" (.getReason terminated)) :reason-container-limitation-memory
+      :default (do
+                 (log/warn "Unable to determine kubernetes state for pod" {:pod-status pod-status
+                                                                           :container-status status})
+                 :unknown))))
+
+(defn handle-status-update
+  "Helper function for calling scheduler/handle-status-update"
+  [kcc mesos-status]
+  (scheduler/write-status-to-datomic datomic/conn
+                                     @(:pool->fenzo-atom kcc)
+                                     mesos-status))
+
+(defn- get-job-container-status
+  "Extract the constainer status for the api/cook-container-name-for-job container"
+  [^V1PodStatus pod-status]
+  (->> pod-status
+       .getContainerStatuses
+       (filter (fn [container-status] (= api/cook-container-name-for-job
+                                         (.getName container-status))))
+       first))
+
+(defn pod-has-just-completed
+  "A pod has completed."
+  [kcc {:keys [synthesized-state pod] :as existing-state-dictionary}]
+  (let [task-id (-> pod .getMetadata .getName)
+        pod-status (.getStatus pod)
+        ^V1ContainerStatus job-container-status (get-job-container-status pod-status)
+        mesos-state (case (:state synthesized-state)
+                      :pod/failed :task-failed
+                      :pod/succeeded :task-finished)
+        status {:task-id {:value task-id}
+                :state mesos-state
+                :reason (container-status->failure-reason pod-status job-container-status)}
+        exit-code (-> job-container-status .getState .getTerminated .getExitCode)]
+    (handle-status-update kcc status)
+    (sandbox/aggregate-exit-code (:exit-code-syncer-state kcc) task-id exit-code)
+    {:expected-state :expected/completed}))
+
+(defn prepare-expected-state-dict-for-logging
+  [expected-state-dict]
+  ".toString on a pod is incredibly large. Make a version thats been elided."
```

Comment:
Flip this with prior line.

Reply:
Are we even calling this, given that the syntax is bad? Can we just remove it?

Reply:
It'll work fine. We just generate the doc string and drop it when executing. (Its dead code, just as if we'd done (defn foo [a b] 1 "2" 3), where 1 and "2" are dead.